### PR TITLE
[feat][broker][PIP-278] Support pluggable topic compaction service - part2

### PR DESCRIPTION
**Pull request by coderzc**
PIP: #20624 

### Motivation & Modifications

* Add `readCompactedEntries` static method
* Migrate `CompactedTopic` and `Compactor` to `TopicCompactionService`
* Fix tests and makes them compatible with new implementations
* For easy reuse by other custom compactors `TwoPhaseCompactor`, abstract `extractIdsAndKeysAndSizeFromBatch` and `rebatchMessage` methods in the `TwoPhaseCompactor`

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java [hunk 1]
```diff
@@ -626,4 +633,57 @@ public void recycle() {
             recyclerHandle.recycle(this);
         }
     }
+
+    public static void readCompactedEntries(TopicCompactionService topicCompactionService, ManagedCursor cursor,
```
Comment (BewareMyPower) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:637]: Why did you make it public? It should be private.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 2]
```diff
@@ -3188,7 +3209,7 @@ public synchronized LongRunningProcessStatus compactionStatus() {
             return LongRunningProcessStatus.forStatus(LongRunningProcessStatus.Status.RUNNING);
         } else {
             try {
-                if (current.join() == COMPACTION_NEVER_RUN) {
+                if (Objects.equals(current.join(), COMPACTION_NEVER_RUN)) {
```
Comment (BewareMyPower) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3212]: There is no need to use `Objects.equals` here. Even if `current.join()` is `null`, `null == COMPACTION_NEVER_RUN` is still a valid expression. The `Objects.equals` method is only used to replace the `x.equals(y)` call in case when `x` is `null`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 3]
```diff
@@ -3167,7 +3188,7 @@ public synchronized void triggerCompaction()
                 currentCompaction = brokerService.pulsar().getStrategicCompactor()
                         .compact(topic, strategicCompactionMap.get(topic));
             } else {
-                currentCompaction = brokerService.pulsar().getCompactor().compact(topic);
+                currentCompaction = topicCompactionService.compact().thenApply(x -> null);
```
Comment (BewareMyPower) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3191]: Why did you add the `thenApply` here? I think `compact()` always completes with `null`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java [hunk 4]
```diff
@@ -347,7 +354,7 @@ protected void readMoreEntries(Consumer consumer) {
                 }
                 havePendingRead = true;
                 if (consumer.readCompacted()) {
-                    topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
```
Comment (BewareMyPower) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:350]: I think `CompactedTopic#asyncReadEntriesOrWait` can be marked as deprecated after this modification.

File: pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java [hunk 5]
```diff
@@ -400,6 +400,16 @@ private static Pair<String, Integer> extractKeyAndSize(RawMessage m) {
...
 
+    protected List<ImmutableTriple<MessageId, String, Integer>> extractIdsAndKeysAndSizeFromBatch(RawMessage msg)
+            throws IOException {
+        return RawBatchConverter.extractIdsAndKeysAndSize(msg);
+    }
+
+    protected Optional<RawMessage> rebatchMessage(RawMessage msg, BiPredicate<String, MessageId> filter)
+            throws IOException {
+        return RawBatchConverter.rebatchMessage(msg, filter);
+    }
```
[the remark this answers is not in the document]
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java:411]: This is easy to Kop extend `TwoPhaseCompactor` and overwriting them.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 6]
```diff
@@ -434,6 +420,37 @@ public AtomicLong getPendingWriteOps() {
         return pendingWriteOps;
     }
 
+    protected CompletableFuture<Void> newTopicCompactionService() {
```
[the remark this answers is not in the document]
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:423]: Yes, it should be private.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java (hunk 1 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:637]: Because it needs to be used by Kop.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java (hunk 4 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:350]: You are right.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 3 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3191]: Because the currentCompaction type is `CompletableFuture<Long>`, however topicCompactionService.compact() return `CompletableFuture<Void>` type.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 2 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3212]: If current.join() is `-0xfebecffeL` then `current.join() != COMPACTION_NEVER_RUN`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 7]
```diff
@@ -210,9 +212,9 @@ public class PersistentTopic extends AbstractTopic implements Topic, AddEntryCal
 
     protected final MessageDeduplication messageDeduplication;
 
-    private static final long COMPACTION_NEVER_RUN = -0xfebecffeL;
+    private static final Long COMPACTION_NEVER_RUN = -0xfebecffeL;
     private CompletableFuture<Long> currentCompaction = CompletableFuture.completedFuture(COMPACTION_NEVER_RUN);
-    private final CompactedTopic compactedTopic;
+    private volatile TopicCompactionService topicCompactionService;
```
Comment (Demogorgon314) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:217]: Why do we need to mark the `TopicCompactionService` as `volatile`?

Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java [hunk 8]
```diff
@@ -1478,25 +1489,14 @@ public synchronized ScheduledExecutorService getCompactorExecutor() {
...
-        }
-        return this.compactor;
-    }
-
     // This method is used for metrics, which is allowed to as null
     // Because it's no operation on the compactor, so let's remove the  synchronized on this method
     // to avoid unnecessary lock competition.
     public Compactor getNullableCompactor() {
-        return this.compactor;
+        if (this.compactionServiceFactory instanceof PulsarCompactionServiceFactory pulsarCompactedServiceFactory) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:1496]: Can you please add some comments here to let other people know why we should instance of the implementation here? :)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java [hunk 9]
```diff
@@ -1914,4 +1914,13 @@ public void shutdownNow() {
     protected BrokerService newBrokerService(PulsarService pulsar) throws Exception {
         return new BrokerService(pulsar, ioEventLoopGroup);
     }
+
+    private CompactionServiceFactory loadCompactionServiceFactory() {
+        String compactionServiceFactoryClassName = config.getCompactionServiceFactoryClassName();
+        var compactionServiceFactory =
+                Reflections.createInstance(compactionServiceFactoryClassName, CompactionServiceFactory.class,
+                        Thread.currentThread().getContextClassLoader());
+        compactionServiceFactory.initialize(this).join();
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:1923]: Well, I am wondering if it's valuable to use get with a timeout to avoid the broker hanging forever in some thread-blocking cases. :)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java (hunk 1 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:637]: How about moving this logic into the `Dispatchers` util class? Because it's a static method, and it's weird for the eco-system to call it in [PersistentDispatcherSingleActiveConsumer.java](https://github.com/apache/pulsar/pull/20718/files?diff=split&w=1#diff-cf3d6c439d725b9688d46adb393e4f36ec0470b201101c1f38e0e90d145cf0b6) :)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java [hunk 10]
```diff
@@ -626,4 +633,57 @@ public void recycle() {
             recyclerHandle.recycle(this);
         }
     }
+
+    public static void readCompactedEntries(TopicCompactionService topicCompactionService, ManagedCursor cursor,
+                                            int numberOfEntriesToRead, boolean isFirstRead,
+                                            AsyncCallbacks.ReadEntriesCallback callback, Consumer consumer) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:639]: You should add some NPE and illegal argument judgement as a public API. :)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java [hunk 11]
```diff
@@ -626,4 +633,57 @@ public void recycle() {
...
+        lastCompactedPositionFuture.thenCompose(lastCompactedPosition -> {
+            if (lastCompactedPosition == null
+                    || readPosition.compareTo(
+                    lastCompactedPosition.getLedgerId(), lastCompactedPosition.getEntryId()) > 0) {
+                cursor.asyncReadEntriesOrWait(numberOfEntriesToRead, callback, readEntriesCtx, PositionImpl.LATEST);
+                return CompletableFuture.completedFuture(null);
+            }
+
+            return topicCompactionService.readCompactedEntries(readPosition, numberOfEntriesToRead)
+                    .thenApply(entries -> {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:661]: Please use ''thenAccept` if there is always a return null.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 12]
```diff
@@ -2517,7 +2530,9 @@ public CompletableFuture<PersistentTopicInternalStats> getInternalStats(boolean
 
     public Optional<CompactedTopicContext> getCompactedTopicContext() {
         try {
-            return ((CompactedTopicImpl) compactedTopic).getCompactedTopicContext();
+            if (topicCompactionService instanceof PulsarTopicCompactionService pulsarCompactedService) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2533]: Does it use for metrics?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 13]
```diff
@@ -481,8 +492,10 @@ public CompletableFuture<Void> unloadSubscription(@Nonnull String subName) {
 
     private PersistentSubscription createPersistentSubscription(String subscriptionName, ManagedCursor cursor,
             boolean replicated, Map<String, String> subscriptionProperties) {
-        Objects.requireNonNull(compactedTopic);
-        if (isCompactionSubscription(subscriptionName)) {
+        Objects.requireNonNull(topicCompactionService);
+        if (isCompactionSubscription(subscriptionName)
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:496]: Do we have any way to avoid using the instance of here?

File: pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopic.java [hunk 14]
```diff
@@ -29,6 +29,7 @@
 public interface CompactedTopic {
     CompletableFuture<CompactedTopicContext> newCompactedLedger(Position p, long compactedLedgerId);
     CompletableFuture<Void> deleteCompactedLedger(long compactedLedgerId);
+    @Deprecated
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopic.java:32]: Please add a comment to help people know which API replace this.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentTopicTest.java [hunk 15]
```diff
@@ -663,6 +665,7 @@ private SubscriptionOption getSubscriptionOption(CommandSubscribe cmd) {
     @Test
     public void testSubscribeUnsubscribe() throws Exception {
         PersistentTopic topic = new PersistentTopic(successTopicName, ledgerMock, brokerService);
+        topic.initialize().join();
```
Comment (mattisonchao) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentTopicTest.java:668]: Q: why add this?

**Review by mattisonchao — APPROVED** [review]
Nice work! 👍🏼 

I left some minor comments.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 12 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2533]: Yes.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java (hunk 1 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java:637]: Good idea.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentTopicTest.java (hunk 15 above)
Comment (coderzc) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentTopicTest.java]: Because `createPersistentSubscription` depends on topicCompactionService object.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java (hunk 9 above)
Comment (Demogorgon314) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:1923]: +1, Why do we let the initialize method return `CompletableFuture`? Do you think we can make it sync?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 16]
```diff
@@ -210,9 +212,9 @@ public class PersistentTopic extends AbstractTopic implements Topic, AddEntryCal
 
     protected final MessageDeduplication messageDeduplication;
 
-    private static final long COMPACTION_NEVER_RUN = -0xfebecffeL;
+    private static final Long COMPACTION_NEVER_RUN = -0xfebecffeL;
     private CompletableFuture<Long> currentCompaction = CompletableFuture.completedFuture(COMPACTION_NEVER_RUN);
-    private final CompactedTopic compactedTopic;
+    private TopicCompactionService topicCompactionService;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:217]: Should we close the compaction service when closing(deleting) the topic? We should provide a close method for the compaction service to make sure the opened resources can be closed when the topic is closed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 17]
```diff
@@ -434,6 +419,32 @@ public AtomicLong getPendingWriteOps() {
         return pendingWriteOps;
     }
 
+    private CompletableFuture<Void> newTopicCompactionService() {
+        CompactionServiceFactory compactionServiceFactory = brokerService.pulsar().getCompactionServiceFactory();
+        return compactionServiceFactory.newTopicCompactionService(topic).thenAccept(topicCompactionService -> {
+            PersistentTopic.this.topicCompactionService = topicCompactionService;
+        });
+    }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:427]: Is it better to move this method to PulsarService?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 18]
```diff
@@ -296,29 +298,11 @@ public PersistentTopic(String topic, ManagedLedger ledger, BrokerService brokerS
...
-                // now
-                subscriptions.get(subscriptionName).deactivateCursor();
-            }
-        }
         this.messageDeduplication = new MessageDeduplication(brokerService.pulsar(), this, ledger);
         if (ledger.getProperties().containsKey(TOPIC_EPOCH_PROPERTY_NAME)) {
             topicEpoch = Optional.of(Long.parseLong(ledger.getProperties().get(TOPIC_EPOCH_PROPERTY_NAME)));
         }
 
-        checkReplicatedSubscriptionControllerState();
```
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:321]: Oh, it has been moved to the `initialize` stage.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 19]
```diff
@@ -481,8 +492,10 @@ public CompletableFuture<Void> unloadSubscription(@Nonnull String subName) {
 
     private PersistentSubscription createPersistentSubscription(String subscriptionName, ManagedCursor cursor,
             boolean replicated, Map<String, String> subscriptionProperties) {
-        Objects.requireNonNull(compactedTopic);
-        if (isCompactionSubscription(subscriptionName)) {
+        Objects.requireNonNull(topicCompactionService);
+        if (isCompactionSubscription(subscriptionName)
+                && topicCompactionService instanceof PulsarTopicCompactionService pulsarTopicCompactionService) {
+            CompactedTopicImpl compactedTopic = pulsarTopicCompactionService.getCompactedTopic();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:498]: What about the compaction service plugin? Will it use the same compaction cursor for maintaining the compaction status?

**Review by codelipenghui — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 16 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:217]: There doesn't seem to be a resource that needs to be closed? , because the resource is shared in the `CompactionserviceFactory`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 19 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:498]: Yes, the compaction cursor (`__compaction`) still be used for maintaining the compaction status.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 20]
```diff
@@ -2070,64 +2070,67 @@ private void getLargestBatchIndexWhenPossible(
...
-        // For a valid position, we read the entry out and parse the batch size from its metadata.
-        CompletableFuture<Entry> entryFuture = new CompletableFuture<>();
-        ml.asyncReadEntry(lastPosition, new AsyncCallbacks.ReadEntryCallback() {
-            @Override
-            public void readEntryComplete(Entry entry, Object ctx) {
-                entryFuture.complete(entry);
+        CompletableFuture<Position> compactionHorizonFuture =
+                persistentTopic.getTopicCompactionService().getLastCompactedPosition();
+
+        compactionHorizonFuture.thenAccept(compactionHorizon -> {
```
[the remark this answers is not in the document]
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2076]: Fixed.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 17 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: done

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 16 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:216]: But maybe the user-defined compaction service will open resources for each topic. For example, if you try to write the compacted to a new topic with a producer (each compaction task will write the compacted to a new topic and then remove the old one). When the original topic is closed, we should also close the producer for writing compacted data? And we can't close the CompactionserviceFactory because the broker is still running, just some topics move to other brokers.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java (hunk 9 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:1926]: @coderzc Maybe you missed this one

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 19 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:493]: Could you point out how the broker creates the __compaction subscription if the topic compaction service is not `PulsarTopicCompactionService`? Or it will be handled by the user?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 21]
```diff
@@ -2070,64 +2070,70 @@ private void getLargestBatchIndexWhenPossible(
...
-                        markDeletePosition != null ? markDeletePosition.getEntryId() : -1));
-            }
+                    writeAndFlush(Commands.newGetLastMessageIdResponse(requestId, lastPosition.getLedgerId(),
+                            lastPosition.getEntryId(), partitionIndex, largestBatchIndex,
+                            markDeletePosition != null ? markDeletePosition.getLedgerId() : -1,
+                            markDeletePosition != null ? markDeletePosition.getEntryId() : -1));
+                }
+            });
+        }).exceptionally(e -> {
+            writeAndFlush(Commands.newError(requestId, ServerError.UnknownError, e.getMessage()));
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2130]: Is it better to use `MetadataError`? I mean the `compactionHorizon` is more like metadata of the topic.

Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2130]: Please add a log to help debug. Thanks!

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java (hunk 9 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:1926]: I think the asynchronous interface is more extensible, and if we support asynchronous initialize in the PulsarService in the future, we don't need to change the interface

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 16 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:216]: Ok, make sense.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 19 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:493]: Other TopicCompactionService also creates the `__compaction` subscription, see: line-496.
In addition, `PulsarCompactorSubscription` only overrides the `aknowledgeMessage` method to early delete historical data, but it is not required the other `TopicCompactionService` can handle it through other way.

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 16 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:216]: I open a new PR to change PIP: https://github.com/apache/pulsar/pull/20786

**Review by Demogorgon314 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 2 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:3209]: Sorry I don't get it. When does `current.join()` return `-0xfebecffeL`?

In addition,

```java
    public static boolean equals(Object a, Object b) {
        return (a == b) || (a != null && a.equals(b));
    }
```

It's equivalent with `a == b` when `a` and `b` are `Long`.

File: pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java:411]: It's not a good idea just for this motivation. We should not apply any guarantee to the protected methods. We don't have legal reasons to prevent others from removing them or just making them private.

It's not friendly for any refactoring. Others might change the method to return a nullable `RawMessage` or even not return a `RawMessage`.

[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java:411]: Let's keep it `protected` for now.

**Review by BewareMyPower — APPROVED** [review]
